pkg/controller/nodeipam/ipam/multinetwork_cloud_cidr_allocator.go

```diff
@@ -0,0 +1,68 @@
+	networkv1 "k8s.io/cloud-provider-gcp/crd/apis/network/v1"
+	"k8s.io/klog/v2"
+)
+
+// PerformMultiNetworkCIDRAllocation allots pod CIDRs for all the networks that a node is connected to. It handles IPv6 only for default-network for now.
+func (ca *cloudCIDRAllocator) PerformMultiNetworkCIDRAllocation(node *v1.Node, interfaces []*compute.NetworkInterface) (defaultNwCIDRs []string, northInterfaces networkv1.NorthInterfacesAnnotation, additionalNodeNetworks networkv1.MultiNetworkAnnotation, err error) {
+	k8sNetworksList, err := ca.networkClient.NetworkingV1().Networks().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("error fetching networks: %v", err)
+	}
+	k8sNetworks := make([]networkv1.Network, 0)
+	// ignore networks that are under deletion.
+	for _, network := range k8sNetworksList.Items {
+		if network.ObjectMeta.DeletionTimestamp.IsZero() {
+			k8sNetworks = append(k8sNetworks, network)
+		}
+	}
+	gkeNwParamsClient := ca.networkClient.NetworkingV1().GKENetworkParamses("")
+	// Fetch the GKENetworkParams for every k8s-network object.
+	// Match the fetched GKENetworkParams object with the interfaces on the node
+	// to build the per-network north-interface and node-network annotations useful for IPAM.
+	for _, network := range k8sNetworks {
+		klog.V(2).Infof("allotting pod cidrs for network %s", network.Name)
+		gnp, err := gkeNwParamsClient.Get(context.TODO(), network.Spec.ParametersRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		for _, inf := range interfaces {
+			if inf.Network != gnp.Spec.VPC || inf.Subnetwork != gnp.Spec.VPCSubnet {
+				continue
+			}
+			klog.V(2).Infof("interface %s matched, proceeding to find a secondary ranges", inf.Name)
+			// TODO: Handle IPv6 in future.
+			secondaryRangeNames := gnp.Spec.PodIPv4Ranges.RangeNames
+			// Each secondary range in a subnet corresponds to a pod-network. AliasIPRanges list on a node interface consists of IP ranges that belong to multiple secondary ranges (pod-networks).
+			// Match the secondary range names of interface and GKENetworkParams and set the right IpCidrRange for current network.
+			rangeNameAliasIPMap := map[string]*compute.AliasIpRange{}
+			for _, ipRange := range inf.AliasIpRanges {
+				rangeNameAliasIPMap[ipRange.SubnetworkRangeName] = ipRange
+			}
+			for _, secondaryRangeName := range secondaryRangeNames {
+				ipRange, ok := rangeNameAliasIPMap[secondaryRangeName]
+				if !ok {
+					continue
+				}
+				klog.V(2).Infof("found an allocatable secondary range for the interface on network")
+				if network.Name == networkv1.DefaultNetworkName {
+					defaultNwCIDRs = append(defaultNwCIDRs, ipRange.IpCidrRange)
+					defaultNwCIDRs = ca.cloud.AccommodateIPV6Addresses(defaultNwCIDRs, inf, node.Spec.ProviderID)
+				} else {
+					northInterfaces = append(northInterfaces, networkv1.NorthInterface{Network: network.Name, IpAddress: inf.NetworkIP})
```

Comment:
You need to do this even in the case where GKENetworkParams does not have any IP range names. This is to handle the high performance network interfaces.

Reply:
Done, adding the northInterface for all non-default networks.